### PR TITLE
[Database] Build Database CommManager

### LINKED ISSUE #71: [Database] Build and Test Database Production Classes
**What is the use case? Please describe.**
For this sprint task, you'll be building and testing all of the classes in the `edu.oakland.production.database` package. Note: This issue is similar to #65 , but is tracked differently as helper and production classes have different deliverable dates.

**Sprint Task Setup**

To be done by @camrongrant1 , @kylepoterek , or @MathewYaldo before the rest of the team starts working on it:

1. Edit this issue message to add a checkbox for each class, and assign it to the correct team member from #60 .

**Prerequisites**

Please makes sure that your test plan from #60 for the class you are writing is closed and merged to main before starting this, to ensure that all of the documentation you need for this is in the correct spot.

**Steps**

To be done by each team member for each class they are developing a test plan for:

1. Each team member should watch the last ~18 minutes of @andrewdimmer 's [Test Driven Development Crash Course](https://youtu.be/eSc0aY4GhPs)
2. One person should create a branch off of main called `define-Name_Of_Class` (ex. the `Demo` classes and tests get build on a branch called `build-Demo`). Everyone else in the programming pair or group should then checkout that branch.
3. Using the information in the video, use test driven development to build all the required classes, stubs, and tests. Remember: you've already defined what you need to build at https://github.com/andrewdimmer/csi-3370-software-project/tree/main/docs/uml-class-diagrams/database !
4. Commit your changes with the message `[Database] Built Name_Of_Class`
5. Push your changes
6. Create a pull request with the name `[Database] Built Name_Of_Class` and the rest of the information like description, tags, etc. For reference, the settings on the right hand side should match this issue, with the exception that the assignees on the PR are just those who are pushing code.
7. Once your code is complete, add @andrewdimmer as a reviewer.

**Tasks to Complete**

- [x] Allocate Team Members to Classes - @camrongrant1 , @kylepoterek , or @MathewYaldo
- [ ] Build DatabaseCommInterface - @Ian-Klaus 
- [x] Build DatabaseCommManager - @Endritz 
- [x] Build DatabseGisInterface - @robertsonbrinker 
- [x] Build DatabaseGisManager - @Jake-wq 
- [x] Build DatabasePersistentStorage - @MathewYaldo @letthatbe 

**Additional context**
Note: if for some reason you need to change a public interface or shared data type, reach out to @andrewdimmer who can help you manage making that BREAKING CHANGE.

Note: if you run into merge conflicts, let @andrewdimmer know and he can help you resolve them.

### PR DESCRIPTION
**Pull request by MathewYaldo**
Working on #71

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (andrewdimmer) [inline .vscode/settings.json]: This file is part of IDE setup, can we remove it from the GitHub repo? Thank you!

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerStub.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package edu.oakland.test.database;
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerStub.java]: You shouldn't need this stub at all for this PR, as you want to test this production class, not test the stub. You'll instead need the `DatabasePersistentStorageStub` to test the `DatabaseCommManager`. This can be deleted from the PR, but keep this file locally in another folder so you can pass it on to the person working on the `DatabaseCommInterface`.

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+import edu.oakland.helper.admin.LocationDataPoint;
+import edu.oakland.helper.admin.TrackData;
+import edu.oakland.production.database.DatabaseCommManagerImplementation;
+import java.lang.IllegalArgumentException;
+import java.time.LocalDateTime;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+@DisplayName("Database CommManager Test")
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java]: For sake of standardization to make it easier to read and search the logs, can we change this to `DatabaseCommManagerImplementation Unit Tests`? Thank you!
```suggestion
@DisplayName("DatabaseCommManagerImplementation Unit Tests")
```

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+public class DatabaseCommManagerTest {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java]: Note: the `DatabaseCommManager` is an Java interface and can't be tested. We want to specify we're testing the implementation. 
```suggestion
public class DatabaseCommManagerImplementationTest {
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageStub.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package edu.oakland.test.database;
+
+import edu.oakland.helper.admin.LocationDataPoint;
+import edu.oakland.helper.admin.TrackData;
+import edu.oakland.production.database.DatabasePersistentStorage;
+
+public class DatabasePersistentStorageStub implements DatabasePersistentStorage {
+  int rfid = 10;
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageStub.java]: Can we make this dynamically configurable? Below is one way to do it, or you could also add a setter.
```suggestion
  private int rfid;
  
  public DatabasePersistentStorageStub(int rfit) {
    this.rfid = rfid;
  }
```

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+  @Test
+  @DisplayName("rfid going in is Rfid coming Out")
+  void rfidInIsRfidOut() {
+
+    int i = 10;
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java]: Can we make this random?
```suggestion
    int i = (int) (Math.random() * 100);
```

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+    DatabasePersistentStorageStub dpss = new DatabasePersistentStorageStub();
+    assertEquals(i, dpss.locateRfidData());
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java]: Our goal here is to test the DatabaseCommManager, so we need to be calling that class in every one of our tests. An example of how to do that for this would be as follows:
```suggestion
    DatabasePersistentStorageStub dpss = new DatabasePersistentStorageStub(i);
    DatabaseCommManager dcm = new DatabaseCommManagerImplementation(dpss);
    assertEquals(i, dcm.passGetRfidRequest());
```

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    float lng = (float) (Math.random() * 360 - 180);
+
+    LocationDataPoint ldp = new LocationDataPoint(lat, lng,
+        LocalDateTime.of((int) (Math.random() * 50 + 1970),
+         (int) (Math.random() * 12 + 1),
+            (int) (Math.random() * 28 + 1),
+             (int) (Math.random() * 24),
+              (int) (Math.random() * 60)));
+
+    ldp = null;
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java]: We shouldn't be setting what we just randomly generated to `null`. `null` is an edge case that should be handled separately, and should not be used as a default test case. Instead, let's pass our randomly generated `LocationDataPoint` to the the `DatabasePersistentStorageStub`.
```suggestion
    dpss.storeLocationDataPoint(ldp);
```

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    int i = (int) Math.random();
+    assertEquals(ldp, dpss.getLocationDataPoint(i));
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java]:
```suggestion
    DatabaseCommManager dcm = new DatabaseCommManagerImplementation(dpss);
    assertEquals(ldp, dcm.passGetLocationDataPointRequest(i));
```

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  @Test
+  @DisplayName("Offset matches TrackData")
+  void offsetMatchTrackData() {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java]: Can you model this test off of the ones above it?

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+    TrackData td = new TrackData(generateRandomLocationDataPointsArray(10));
+
+    int i = (int) (Math.random() * 7);
+    assertEquals(null, dpss.getTrackData(i));
+  }
+
+  @Test
+  @DisplayName("TrackData In is same as TrackData Out")
+  void trackDataInIsTrackDataOut() {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java]: Can you model this test off of the ones above it?

**Review by andrewdimmer — CHANGES_REQUESTED** [review]
I've made a number of suggestions to help you get started fixing the testing. If you want to go through and accept my suggestions directly, feel free to. You can also accept as many or as few as you'd like in a batch from https://github.com/andrewdimmer/csi-3370-software-project/pull/136/files, by adding each suggestion you want to add to a batch, then committing the whole batch all at once from the green commit button at the top of the page.

Note that my suggestions are just fixing the lines they are on; you'll likely need to fix additional things like imports or references on other lines to get everything to compile again. They should provide a baseline to model how to do the first two tests though, so you can apply the same techniques to the last two. Let me know if you have any questions, and I'd be happy to provide more info if you need it!

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+import edu.oakland.helper.admin.LocationDataPoint;
+import edu.oakland.helper.admin.TrackData;
+import edu.oakland.production.database.DatabaseCommManagerImplementation;
+import java.lang.IllegalArgumentException;
+import java.time.LocalDateTime;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+@DisplayName("DatabaseCommManagerImplementation Unit Tests")
+public class DatabaseCommManagerImplementationTest {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerTest.java]: One more note I missed earlier:
```suggestion
public class DatabaseCommManagerImplementationTests {
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageStub.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import edu.oakland.production.database.DatabasePersistentStorage;
+
+public class DatabasePersistentStorageStub implements DatabasePersistentStorage {
+  private int rfid;
+  private TrackData trackData;
+  
+  public DatabasePersistentStorageStub(int rfid, TrackData trackdata) {
+    this.rfid = rfid;
+    this.trackData = trackdata;
+  }
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageStub.java]: Let's try just doing an RFID for the constructor, as that is easy to set, and TrackData is not.
```suggestion
  public DatabasePersistentStorageStub(int rfid) {
    this.rfid = rfid;
  }
  
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageStub.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+  public void storeLocationDataPoint(LocationDataPoint locationDataPoint) {
+
+  }
+
+  public TrackData getTrackData(int offset) {
+    return trackData;
+  }
+
+  public void storeTrackData(TrackData trackData) {
+
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageStub.java]: This should set up the setter:
```suggestion
    this.trackData = trackData;
```

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerImplementationTest.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+    assertEquals(j, dcm.passGetTrackDataRequest(j));
+  }
+
+  @Test
+  @DisplayName("TrackData In is same as TrackData Out")
+  void trackDataInIsTrackDataOut() {
+    int i = (int) (Math.random() * 100);
+    TrackData td = new TrackData(generateRandomLocationDataPointsArray(10));
+    DatabasePersistentStorageStub dpss = new DatabasePersistentStorageStub(i, td);
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerImplementationTest.java]: Modified the constructor (see the stub file):
```suggestion
    DatabasePersistentStorageStub dpss = new DatabasePersistentStorageStub(i);
```

**Review by andrewdimmer — COMMENTED** [review]
The issue is that you are testing the setter, so it needs to be implemented in the stub. Does that work?

**Review by andrewdimmer — APPROVED** [review]
Looks good to me, all set to merge!

**Review by camrongrant1 — APPROVED** [review]
